File: src/reaverbots/SeededRandom.js
const UINT32_RANGE = 0x100000000;

export function hashSeed(value) {
  const text = typeof value === 'string' ? value : JSON.stringify(value ?? 'reaverbot');
  let hash = 0x811c9dc5;

  for (let index = 0; index < text.length; index += 1) {
    hash ^= text.charCodeAt(index);
    hash = Math.imul(hash, 0x01000193);
  }

  hash += hash << 13;
  hash ^= hash >>> 7;
  hash += hash << 3;
  hash ^= hash >>> 17;
  hash += hash << 5;
  return hash >>> 0;
}

export class SeededRandom {
  constructor(seed = 'reaverbot') {
    this.seed = hashSeed(seed);
    this.state = this.seed || 0x6d2b79f5;
  }

  next() {
    this.state = (this.state + 0x6d2b79f5) >>> 0;
    let value = this.state;
    value = Math.imul(value ^ (value >>> 15), value | 1);
    value ^= value + Math.imul(value ^ (value >>> 7), value | 61);
    return ((value ^ (value >>> 14)) >>> 0) / UINT32_RANGE;
  }

  float(min = 0, max = 1) {
    return min + (max - min) * this.next();
  }

  int(min, max) {
    const lower = Math.ceil(Math.min(min, max));
    const upper = Math.floor(Math.max(min, max));
    return lower + Math.floor(this.next() * (upper - lower + 1));
  }

  chance(probability) {
    return this.next() < probability;
  }

  pick(values) {
    if (!values?.length) {
      return null;
    }

    return values[Math.floor(this.next() * values.length)];
  }

  weighted(entries, fallback = null) {
    const valid = entries.filter((entry) => Number(entry?.weight ?? entry?.[1]) > 0);
    const total = valid.reduce((sum, entry) => sum + Number(entry.weight ?? entry[1]), 0);

    if (total <= 0) {
      return fallback;
    }

    let roll = this.next() * total;
    for (const entry of valid) {
      roll -= Number(entry.weight ?? entry[1]);
      if (roll <= 0) {
        return entry.value ?? entry[0];
      }
    }

    const last = valid.at(-1);
    return last?.value ?? last?.[0] ?? fallback;
  }

  fork(label) {
    return new SeededRandom(`${this.seed}:${String(label)}`);
  }
}

export function createSeededRandom(seed) {
  const random = new SeededRandom(seed);
  return () => random.next();
}

File: unity/RuinCrawler/Assets/RuinCrawler/Runtime/Determinism/SeededRandom.cs
using System;
using System.Collections.Generic;

namespace RuinCrawler.Port.Determinism
{
    /// <summary>
    /// Bit-for-bit port of src/reaverbots/SeededRandom.js for string seeds.
    /// Keep this numeric kernel shared by Unity generation and simulation.
    /// </summary>
    public sealed class SeededRandom
    {
        private const uint DefaultNonZeroState = 0x6d2b79f5u;
        private const double Uint32Range = 4294967296.0;

        public uint Seed { get; }
        public uint State { get; private set; }

        public SeededRandom(string seed = "reaverbot")
        {
            Seed = HashSeed(seed ?? "reaverbot");
            State = Seed == 0 ? DefaultNonZeroState : Seed;
        }

        public static uint HashSeed(string value)
        {
            string text = value ?? "reaverbot";
            uint hash = 0x811c9dc5u;

            unchecked
            {
                for (int index = 0; index < text.Length; index += 1)
                {
                    hash ^= text[index];
                    hash *= 0x01000193u;
                }

                hash += hash << 13;
                hash ^= hash >> 7;
                hash += hash << 3;
                hash ^= hash >> 17;
                hash += hash << 5;
            }

            return hash;
        }

        public double NextDouble()
        {
            uint value;
            unchecked
            {
                State += DefaultNonZeroState;
                value = State;
                value = (value ^ (value >> 15)) * (value | 1u);
                value ^= value + ((value ^ (value >> 7)) * (value | 61u));
                value ^= value >> 14;
            }

            return value / Uint32Range;
        }

        public float Range(float minimum = 0f, float maximum = 1f)
        {
            return minimum + (maximum - minimum) * (float)NextDouble();
        }

        public int RangeInclusive(int minimum, int maximum)
        {
            int lower = Math.Min(minimum, maximum);
            int upper = Math.Max(minimum, maximum);
            return lower + (int)Math.Floor(NextDouble() * (upper - lower + 1d));
        }

        public bool Chance(double probability)
        {
            return NextDouble() < probability;
        }

        public T Pick<T>(IReadOnlyList<T> values)
        {
            if (values == null || values.Count == 0)
            {
                return default;
            }

            return values[(int)Math.Floor(NextDouble() * values.Count)];
        }

        public SeededRandom Fork(string label)
        {
            return new SeededRandom($"{Seed}:{label ?? string.Empty}");
        }
    }
}
